tests/mechanic/supplier_test.py

```diff
@@ -215,6 +221,39 @@ def add_es_artifact(binaries):
         self.assertFalse(cached_supplier.cached)
         self.assertIn("elasticsearch", binaries)
 
+    @mock.patch("os.path.exists")
+    @mock.patch("esrally.mechanic.supplier.ElasticsearchSourceSupplier")
+    def test_uses_already_cached_artifact(self, es, path_exists):
+        # assume that the artifact is already cached
+        path_exists.return_value = True
+        renderer = supplier.TemplateRenderer(version="abc123", os_name="linux", arch="x86_64")
+
+        dist_cfg = {
+            "runtime.jdk.bundled": "true",
+            "jdk.bundled.release_url": "https://elstc.co/elasticsearch-{{VERSION}}-{{OSNAME}}-{{ARCH}}.tar.gz"
+        }
+        file_resolver = supplier.ElasticsearchFileNameResolver(
+            distribution_config=dist_cfg,
+            template_renderer=renderer
+        )
+        cached_supplier = supplier.CachedSourceSupplier(distributions_root="/tmp",
+                                                        source_supplier=es,
+                                                        file_resolver=file_resolver)
+
+        cached_supplier.fetch()
+        cached_supplier.prepare()
+
+        binaries = {}
+
+        cached_supplier.add(binaries)
+
+        self.assertEqual(0, es.fetch.call_count)
+        self.assertEqual(0, es.prepare.call_count)
+        self.assertEqual(0, es.add.call_count)
+        self.assertTrue(cached_supplier.cached)
+        self.assertIn("elasticsearch", binaries)
+        self.assertEqual("/tmp/elasticsearch-abc123-linux-x86_64.tar.gz", binaries["elasticsearch"])
```

Comment:
Do you think its worth testing this assert in the `test_does_not_cache_when_no_revision` test above? 

```
        self.assertEqual("/path/to/artifact.tar.gz", binaries["elasticsearch"])
```

Reply:
Thanks for that; makes sense to me. I've pushed the additional assert in https://github.com/elastic/rally/pull/1028/commits/504d5c30256c3d1f8a247cf1abc147daeb81ffb0.